service/controller/v24/resource/cpf/create.go

```diff
@@ -0,0 +1,127 @@
+package cpf
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+
+	"github.com/giantswarm/aws-operator/service/controller/v24/adapter"
+	"github.com/giantswarm/aws-operator/service/controller/v24/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/v24/key"
+	"github.com/giantswarm/aws-operator/service/controller/v24/templates"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		if cc.Status.Cluster.VPCPeeringConnectionID == "" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find the VPC Peering Connection ID in the controller context")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			resourcecanceledcontext.SetCanceled(ctx)
+			return nil
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding the tenant cluster's control plane finalizer cloud formation stack")
+
+		i := &cloudformation.DescribeStacksInput{
+			StackName: aws.String(key.MainHostPreStackName(cr)),
+		}
+
+		_, err = r.cloudFormation.DescribeStacks(i)
+		if IsNotExists(err) {
+			// fall through
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "found the tenant cluster's control plane finalizer cloud formation stack already exists")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+			return nil
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "did not find the tenant cluster's control plane finalizer cloud formation stack")
+	}
+
+	var templateBody string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "computing the template of the tenant cluster's control plane finalizer cloud formation stack")
+
+		var cpf *adapter.CPF
+		{
+			c := adapter.CPFConfig{
+				RouteTable: r.routeTable,
+
+				BaseDomain:                 key.BaseDomain(cr),
+				ClusterID:                  key.ClusterID(cr),
+				EncrypterBackend:           r.encrypterBackend,
+				GuestHostedZoneNameServers: cc.Status.Cluster.HostedZoneNameServers,
+				Route53Enabled:             r.route53Enabled,
+			}
+
+			cpf, err = adapter.NewCPF(c)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			err = cpf.Boot(ctx)
```

Comment:
I feel like the boot mechanism is not perfect here. It would be nice to only have the data structures configured so that the templates could be rendered without any dependencies involved. The situation improves further but we are still not there.